// @adobe/alloy/libEs5/components/Personalization/createClickStorage.js
"use strict";

exports.default = void 0;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var metasToArray = function metasToArray(metas) {
  return Object.keys(metas).map(function (key) {
    return _objectSpread({
      id: key
    }, metas[key]);
  });
};
var _default = function _default() {
  var clickStorage = {};
  var storeClickMetrics = function storeClickMetrics(_ref) {
    var selector = _ref.selector,
      _ref$meta = _ref.meta,
      id = _ref$meta.id,
      scope = _ref$meta.scope,
      scopeDetails = _ref$meta.scopeDetails,
      trackingLabel = _ref$meta.trackingLabel,
      scopeType = _ref$meta.scopeType;
    if (!clickStorage[selector]) {
      clickStorage[selector] = {};
    }
    clickStorage[selector][id] = {
      scope: scope,
      scopeDetails: scopeDetails,
      trackingLabel: trackingLabel,
      scopeType: scopeType
    };
  };
  var getClickSelectors = function getClickSelectors() {
    return Object.keys(clickStorage);
  };
  var getClickMetasBySelector = function getClickMetasBySelector(selector) {
    var metas = clickStorage[selector];
    if (!metas) {
      return {};
    }
    return metasToArray(clickStorage[selector]);
  };
  return {
    storeClickMetrics: storeClickMetrics,
    getClickSelectors: getClickSelectors,
    getClickMetasBySelector: getClickMetasBySelector
  };
};
exports.default = _default;
