// @adobe/alloy/libEs5/core/createEventManager.js
"use strict";

exports.default = void 0;
var _pageWideScope = require("../constants/pageWideScope");
var _utils = require("../utils");
var _request = require("../utils/request");
var _excluded = ["edgeConfigOverrides"];
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var EVENT_CANCELLATION_MESSAGE = "Event was canceled because the onBeforeEventSend callback returned false.";
var _default = function _default(_ref) {
  var config = _ref.config,
    logger = _ref.logger,
    lifecycle = _ref.lifecycle,
    consent = _ref.consent,
    createEvent = _ref.createEvent,
    createDataCollectionRequestPayload = _ref.createDataCollectionRequestPayload,
    createDataCollectionRequest = _ref.createDataCollectionRequest,
    sendEdgeNetworkRequest = _ref.sendEdgeNetworkRequest,
    _applyResponse = _ref.applyResponse;
  var onBeforeEventSend = config.onBeforeEventSend,
    globalConfigOverrides = config.edgeConfigOverrides;
  return {
    createEvent: createEvent,
    /**
     * Sends an event. This includes running the event and payload through the
     * appropriate lifecycle hooks, sending the request to the server, and
     * handling the response.
     * @param {Object} event This will be JSON stringified and used inside the
     * request payload.
     * @param {Object} [options] Options to pass on to the onBeforeEvent
     * lifecycle method
     * @param {Object} [options.edgeConfigOverrides] Settings that take
     * precedence over the global datastream configuration, including which
     * datastream to use.
     * @returns {*}
     */
    sendEvent: function sendEvent(event) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var localConfigOverrides = options.edgeConfigOverrides,
        otherOptions = _objectWithoutProperties(options, _excluded);
      var requestParams = (0, _request.createRequestParams)({
        payload: createDataCollectionRequestPayload(),
        localConfigOverrides: localConfigOverrides,
        globalConfigOverrides: globalConfigOverrides
      });
      var request = createDataCollectionRequest(requestParams);
      var onResponseCallbackAggregator = (0, _utils.createCallbackAggregator)();
      var onRequestFailureCallbackAggregator = (0, _utils.createCallbackAggregator)();
      return lifecycle.onBeforeEvent(_objectSpread(_objectSpread({}, otherOptions), {}, {
        event: event,
        onResponse: onResponseCallbackAggregator.add,
        onRequestFailure: onRequestFailureCallbackAggregator.add
      })).then(function () {
        requestParams.payload.addEvent(event);
        return consent.awaitConsent();
      }).then(function () {
        try {
          // NOTE: this calls onBeforeEventSend callback (if configured)
          event.finalize(onBeforeEventSend);
        } catch (error) {
          var throwError = function throwError() {
            throw error;
          };
          onRequestFailureCallbackAggregator.add(lifecycle.onRequestFailure);
          return onRequestFailureCallbackAggregator.call({
            error: error
          }).then(throwError, throwError);
        }

        // if the callback returns false, the event should not be sent
        if (!event.shouldSend()) {
          onRequestFailureCallbackAggregator.add(lifecycle.onRequestFailure);
          logger.info(EVENT_CANCELLATION_MESSAGE);
          var error = new Error(EVENT_CANCELLATION_MESSAGE);
          return onRequestFailureCallbackAggregator.call({
            error: error
          }).then(function () {
            // Ensure the promise gets resolved with undefined instead
            // of an array of return values from the callbacks.
          });
        }
        return sendEdgeNetworkRequest({
          request: request,
          runOnResponseCallbacks: onResponseCallbackAggregator.call,
          runOnRequestFailureCallbacks: onRequestFailureCallbackAggregator.call
        });
      });
    },
    applyResponse: function applyResponse(event) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var _options$renderDecisi = options.renderDecisions,
        renderDecisions = _options$renderDecisi === void 0 ? false : _options$renderDecisi,
        _options$decisionCont = options.decisionContext,
        decisionContext = _options$decisionCont === void 0 ? {} : _options$decisionCont,
        _options$responseHead = options.responseHeaders,
        responseHeaders = _options$responseHead === void 0 ? {} : _options$responseHead,
        _options$responseBody = options.responseBody,
        responseBody = _options$responseBody === void 0 ? {
          handle: []
        } : _options$responseBody,
        personalization = options.personalization;
      var payload = createDataCollectionRequestPayload();
      var request = createDataCollectionRequest({
        payload: payload
      });
      var onResponseCallbackAggregator = (0, _utils.createCallbackAggregator)();
      return lifecycle.onBeforeEvent({
        event: event,
        renderDecisions: renderDecisions,
        decisionContext: decisionContext,
        decisionScopes: [_pageWideScope.default],
        personalization: personalization,
        onResponse: onResponseCallbackAggregator.add,
        onRequestFailure: _utils.noop
      }).then(function () {
        payload.addEvent(event);
        return _applyResponse({
          request: request,
          responseHeaders: responseHeaders,
          responseBody: responseBody,
          runOnResponseCallbacks: onResponseCallbackAggregator.call
        });
      });
    }
  };
};
exports.default = _default;
