// @adobe/alloy/libEs5/components/Personalization/handlers/createProcessDomAction.js
"use strict";

exports.default = void 0;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
/*
Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var _default = function _default(_ref) {
  var modules = _ref.modules,
    logger = _ref.logger,
    storeClickMetrics = _ref.storeClickMetrics;
  return function (item) {
    var _ref2 = item.getData() || {},
      type = _ref2.type,
      selector = _ref2.selector;
    if (!type) {
      logger.warn("Invalid DOM action data: missing type.", item.getData());
      return {
        setRenderAttempted: false,
        includeInNotification: false
      };
    }
    if (type === "click") {
      if (!selector) {
        logger.warn("Invalid DOM action data: missing selector.", item.getData());
        return {
          setRenderAttempted: false,
          includeInNotification: false
        };
      }
      storeClickMetrics({
        selector: selector,
        meta: _objectSpread(_objectSpread({}, item.getProposition().getNotification()), {}, {
          trackingLabel: item.getTrackingLabel(),
          scopeType: item.getProposition().getScopeType()
        })
      });
      return {
        setRenderAttempted: true,
        includeInNotification: false
      };
    }
    if (!modules[type]) {
      logger.warn("Invalid DOM action data: unknown type.", item.getData());
      return {
        setRenderAttempted: false,
        includeInNotification: false
      };
    }
    return {
      render: function render() {
        return modules[type](item.getData());
      },
      setRenderAttempted: true,
      includeInNotification: true
    };
  };
};
exports.default = _default;
