// @adobe/alloy/libEs5/components/DecisioningEngine/createEvaluableRulesetPayload.js
"use strict";

exports.default = void 0;
var _aepRulesEngine = require("@adobe/aep-rules-engine");
var _schema = require("../../constants/schema");
var _eventType = require("../../constants/eventType");
var _utils = require("./utils");
var _flattenArray = require("../../utils/flattenArray");
var _createConsequenceAdapter = require("./createConsequenceAdapter");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /*
                                                                                                                                                                                                                  Copyright 2023 Adobe. All rights reserved.
                                                                                                                                                                                                                  This file is licensed to you under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                  you may not use this file except in compliance with the License. You may obtain a copy
                                                                                                                                                                                                                  of the License at http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  Unless required by applicable law or agreed to in writing, software distributed under
                                                                                                                                                                                                                  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
                                                                                                                                                                                                                  OF ANY KIND, either express or implied. See the License for the specific language
                                                                                                                                                                                                                  governing permissions and limitations under the License.
                                                                                                                                                                                                                  */
var isRulesetItem = function isRulesetItem(item) {
  var schema = item.schema,
    data = item.data;
  if (schema === _schema.RULESET_ITEM) {
    return true;
  }
  if (schema !== _schema.JSON_CONTENT_ITEM) {
    return false;
  }
  try {
    var content = typeof data.content === "string" ? JSON.parse(data.content) : data.content;
    return content && Object.prototype.hasOwnProperty.call(content, "version") && Object.prototype.hasOwnProperty.call(content, "rules");
  } catch (error) {
    return false;
  }
};
var _default = function _default(payload, eventRegistry, decisionHistory) {
  var consequenceAdapter = (0, _createConsequenceAdapter.default)();
  var activityId = (0, _utils.getActivityId)(payload);
  var items = [];
  var addItem = function addItem(item) {
    var _item$data = item.data,
      data = _item$data === void 0 ? {} : _item$data,
      schema = item.schema;
    var content = schema === _schema.RULESET_ITEM ? data : data.content;
    if (!content) {
      return;
    }
    items.push((0, _aepRulesEngine.default)(typeof content === "string" ? JSON.parse(content) : content));
  };
  var evaluate = function evaluate(context) {
    var displayEvent = eventRegistry.getEvent(_eventType.DISPLAY, activityId);
    var displayedDate = displayEvent ? displayEvent.firstTimestamp : undefined;
    var qualifyingItems = (0, _flattenArray.default)(items.map(function (item) {
      return item.execute(context);
    })).map(consequenceAdapter).map(function (item) {
      var _ref = decisionHistory.recordQualified(activityId) || {},
        qualifiedDate = _ref.firstTimestamp;
      return _objectSpread(_objectSpread({}, item), {}, {
        data: _objectSpread(_objectSpread({}, item.data), {}, {
          qualifiedDate: qualifiedDate,
          displayedDate: displayedDate
        })
      });
    });
    return _objectSpread(_objectSpread({}, payload), {}, {
      items: qualifyingItems
    });
  };
  if (Array.isArray(payload.items)) {
    payload.items.filter(isRulesetItem).forEach(addItem);
  }
  return {
    evaluate: evaluate,
    isEvaluable: items.length > 0
  };
};
exports.default = _default;
