// @adobe/alloy/libEs5/components/DecisioningEngine/createSubscribeRulesetItems.js
"use strict";

exports.default = void 0;
var _validation = require("../../utils/validation");
var _createSubscription = require("../../utils/createSubscription");
var _utils = require("../../utils");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /*
                                                                                                                                                                                                                  Copyright 2023 Adobe. All rights reserved.
                                                                                                                                                                                                                  This file is licensed to you under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                  you may not use this file except in compliance with the License. You may obtain a copy
                                                                                                                                                                                                                  of the License at http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  Unless required by applicable law or agreed to in writing, software distributed under
                                                                                                                                                                                                                  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
                                                                                                                                                                                                                  OF ANY KIND, either express or implied. See the License for the specific language
                                                                                                                                                                                                                  governing permissions and limitations under the License.
                                                                                                                                                                                                                  */
var validateOptions = function validateOptions(_ref) {
  var options = _ref.options;
  var validator = (0, _validation.objectOf)({
    surfaces: (0, _validation.arrayOf)((0, _validation.string)()).uniqueItems(),
    schemas: (0, _validation.arrayOf)((0, _validation.string)()).uniqueItems(),
    callback: (0, _validation.callback)().required()
  }).noUnknownFields();
  return validator(options);
};
var emissionPreprocessor = function emissionPreprocessor(params, propositions) {
  var surfacesFilter = params.surfacesFilter,
    schemasFilter = params.schemasFilter;
  var result = {
    propositions: propositions.filter(function (payload) {
      return surfacesFilter ? (0, _utils.includes)(surfacesFilter, payload.scope) : true;
    }).map(function (payload) {
      var _payload$items = payload.items,
        items = _payload$items === void 0 ? [] : _payload$items;
      return _objectSpread(_objectSpread({}, payload), {}, {
        items: items.filter(function (item) {
          return schemasFilter ? (0, _utils.includes)(schemasFilter, item.schema) : true;
        })
      });
    }).filter(function (payload) {
      return payload.items.length > 0;
    })
  };
  return [result];
};
var emissionCondition = function emissionCondition(params, result) {
  return result.propositions.length > 0;
};
var _default = function _default() {
  var subscription = (0, _createSubscription.default)();
  subscription.setEmissionPreprocessor(emissionPreprocessor);
  subscription.setEmissionCondition(emissionCondition);
  var run = function run(_ref2) {
    var surfaces = _ref2.surfaces,
      schemas = _ref2.schemas,
      callback = _ref2.callback;
    var unsubscribe = subscription.add(callback, {
      surfacesFilter: surfaces instanceof Array ? surfaces : undefined,
      schemasFilter: schemas instanceof Array ? schemas : undefined
    });
    return Promise.resolve({
      unsubscribe: unsubscribe
    });
  };
  var optionsValidator = function optionsValidator(options) {
    return validateOptions({
      options: options
    });
  };
  var refresh = function refresh(propositions) {
    if (!subscription.hasSubscriptions()) {
      return;
    }
    subscription.emit(propositions);
  };
  return {
    refresh: refresh,
    command: {
      optionsValidator: optionsValidator,
      run: run
    }
  };
};
exports.default = _default;
