// @adobe/alloy/libEs5/components/DecisioningEngine/utils.js
"use strict";

exports.hasExperienceData = exports.getExpirationDate = exports.getDecisionProvider = exports.getActivityId = exports.createSaveStorage = exports.createRestoreStorage = exports.createInMemoryStorage = exports.clearLocalStorage = void 0;
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
/*
Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var createRestoreStorage = function createRestoreStorage(storage, storageKey) {
  return function (defaultValue) {
    var stored = storage.getItem(storageKey);
    if (!stored) {
      return defaultValue;
    }
    try {
      return JSON.parse(stored);
    } catch (e) {
      return defaultValue;
    }
  };
};
exports.createRestoreStorage = createRestoreStorage;
var createSaveStorage = function createSaveStorage(storage, storageKey) {
  var prepareFn = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function (value) {
    return value;
  };
  return function (value) {
    storage.setItem(storageKey, JSON.stringify(prepareFn(value)));
  };
};
exports.createSaveStorage = createSaveStorage;
var getExpirationDate = function getExpirationDate(retentionPeriod) {
  var expirationDate = new Date();
  expirationDate.setDate(expirationDate.getDate() - retentionPeriod);
  return expirationDate;
};
exports.getExpirationDate = getExpirationDate;
var getActivityId = function getActivityId(proposition) {
  var _proposition$scopeDet = proposition.scopeDetails,
    scopeDetails = _proposition$scopeDet === void 0 ? {} : _proposition$scopeDet;
  var _scopeDetails$activit = scopeDetails.activity,
    activity = _scopeDetails$activit === void 0 ? {} : _scopeDetails$activit;
  var id = activity.id;
  return id;
};
exports.getActivityId = getActivityId;
var createInMemoryStorage = function createInMemoryStorage() {
  var inMemoryStorage = {};
  return {
    getItem: function getItem(key) {
      return key in inMemoryStorage ? inMemoryStorage[key] : null;
    },
    setItem: function setItem(key, value) {
      inMemoryStorage[key] = value;
    }
  };
};
exports.createInMemoryStorage = createInMemoryStorage;
var clearLocalStorage = function clearLocalStorage(storage) {
  storage.clear();
};
exports.clearLocalStorage = clearLocalStorage;
var hasExperienceData = function hasExperienceData(xdm) {
  var _xdm$eventType = xdm.eventType,
    eventType = _xdm$eventType === void 0 ? "" : _xdm$eventType,
    _experience = xdm._experience;
  if (!eventType || eventType === "") {
    return false;
  }
  if (!_experience || _typeof(_experience) !== "object") {
    return false;
  }
  return true;
};
exports.hasExperienceData = hasExperienceData;
var getDecisionProvider = function getDecisionProvider(proposition) {
  var _proposition$scopeDet2 = proposition.scopeDetails,
    scopeDetails = _proposition$scopeDet2 === void 0 ? {} : _proposition$scopeDet2;
  var decisionProvider = scopeDetails.decisionProvider;
  return decisionProvider;
};
exports.getDecisionProvider = getDecisionProvider;
