// @adobe/alloy/libEs5/components/DataCollector/validateUserEventOptions.js
"use strict";

exports.default = void 0;
var _validation = require("../../utils/validation");
var _utils = require("../../utils");
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
/**
 * Verifies user provided event options.
 * @param {*} options The user event options to validate
 * @returns {*} Validated options
 */
var _default = function _default(_ref) {
  var options = _ref.options;
  var eventOptionsValidator = (0, _validation.objectOf)({
    type: (0, _validation.string)(),
    xdm: (0, _validation.objectOf)({
      eventType: (0, _validation.string)(),
      identityMap: _utils.validateIdentityMap
    }),
    data: (0, _validation.objectOf)({}),
    documentUnloading: (0, _validation.boolean)(),
    renderDecisions: (0, _validation.boolean)(),
    decisionScopes: (0, _validation.arrayOf)((0, _validation.string)()).uniqueItems(),
    personalization: (0, _validation.objectOf)({
      decisionScopes: (0, _validation.arrayOf)((0, _validation.string)()).uniqueItems(),
      surfaces: (0, _validation.arrayOf)((0, _validation.string)()).uniqueItems(),
      sendDisplayEvent: (0, _validation.boolean)().default(true),
      includeRenderedPropositions: (0, _validation.boolean)().default(false),
      defaultPersonalizationEnabled: (0, _validation.boolean)(),
      decisionContext: (0, _validation.objectOf)({})
    }).default({
      sendDisplayEvent: true
    }),
    datasetId: (0, _validation.string)(),
    mergeId: (0, _validation.string)(),
    edgeConfigOverrides: _utils.validateConfigOverride,
    initializePersonalization: (0, _validation.boolean)()
  }).required().noUnknownFields();
  return eventOptionsValidator(options);
};
exports.default = _default;
