// @adobe/alloy/libEs5/components/Personalization/in-app-message-actions/utils.js
"use strict";

exports.removeElementById = exports.parseAnchor = void 0;
var _utils = require("../../../utils");
var _dom = require("../../../utils/dom");
/*
Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var removeElementById = function removeElementById(id) {
  var element = (0, _dom.selectNodes)("#" + id, document);
  if (element && element.length > 0) {
    (0, _dom.removeNode)(element[0]);
  }
};
exports.removeElementById = removeElementById;
var parseAnchor = function parseAnchor(anchor) {
  var nothing = {};
  if (!anchor || anchor.tagName.toLowerCase() !== "a") {
    return nothing;
  }
  var href = anchor.href;
  if (!href || !(0, _utils.startsWith)(href, "adbinapp://")) {
    return nothing;
  }
  var hrefParts = href.split("?");
  var action = hrefParts[0].split("://")[1];
  var label = anchor.innerText;
  var uuid = anchor.getAttribute("data-uuid") || "";
  var interaction;
  var link;
  if ((0, _utils.isNonEmptyArray)(hrefParts)) {
    var queryParams = _utils.queryString.parse(hrefParts[1]);
    interaction = queryParams.interaction || "";
    link = decodeURIComponent(queryParams.link || "");
  }
  return {
    action: action,
    interaction: interaction,
    link: link,
    label: label,
    uuid: uuid
  };
};
exports.parseAnchor = parseAnchor;
